AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt

```diff
@@ -1689,88 +1693,98 @@ open class DeckPicker :
-                        getColUnsafe.decks.selected(),
+                        did,
                         true
                     )
                     showDialogFragment(d)
                 }
             }
 
-            // Check if we need to update the fragment or update the deck list. The same checks
-            // are required for all snackbars below.
+        fun showEmptyDeckSnackbar() = showSnackbar(R.string.empty_deck) {
+            addCallback(mSnackbarShowHideCallback)
+            setAction(R.string.menu_add) { addNote() }
+        }
+
+        fun showCustomStudySnackbar() = showSnackbar(R.string.studyoptions_empty_schedule) {
+            addCallback(mSnackbarShowHideCallback)
+            setAction(R.string.custom_study) {
+                val d = mCustomStudyDialogFactory.newCustomStudyDialog().withArguments(
+                    CustomStudyDialog.ContextMenuConfiguration.EMPTY_SCHEDULE,
+                    did,
+                    true
+                )
+                showDialogFragment(d)
+            }
+        }
+
+        /** Check if we need to update the fragment or update the deck list */
+        fun updateUi() {
             if (fragmented) {
                 // Tablets must always show the study options that corresponds to the current deck,
                 // regardless of whether the deck is currently reviewable or not.
-                openStudyOptions(false)
+                openStudyOptions(withDeckOptions = false)
             } else {
                 // On phones, we update the deck list to ensure the currently selected deck is
                 // highlighted correctly.
                 updateDeckList()
             }
-        } else if (getColUnsafe.decks.isDyn(did)) {
-            // Go to the study options screen if filtered deck with no cards to study
-            openStudyOptions(false)
-        } else if (deckDueTreeNode.children.isEmpty() && getColUnsafe.isEmptyDeck(did)) {
-            // If the deck is empty and has no children then show a message saying it's empty
-            showSnackbar(R.string.empty_deck) {
-                addCallback(mSnackbarShowHideCallback)
-                setAction(R.string.menu_add) { addNote() }
+        }
+
+        // Reduce the number of `withCol` calls in the method: access both `config` and `decks`
+        val mustUpdateScheduler = withCol {
+            if ((config.get("schedVer") ?: 1L) == 1L) {
+                return@withCol true
             }
+            decks.select(did)
+            // Also forget the last deck used by the Browser
+            CardBrowser.clearLastDeckId()
+            mFocusedDeck = did
+            false
+        }
 
-            if (fragmented) {
-                openStudyOptions(false)
-            } else {
-                updateDeckList()
+        if (mustUpdateScheduler) {
+            promptUserToUpdateScheduler()
+            return
+        }
```

Comment:
Here I think that it is easier to read if the `config` and `deck` calls are separated. As I read, it should be a guard clause like: `if the scheduler is v1 → Prompt to update and return early (i.e. don't do anything else)`

```suggestion
        if (withCol { ((config.get("schedVer") ?: 1L) == 1L) }) {
            promptUserToUpdateScheduler()
            return
        }
        
        withCol { decks.select(did) }
        // Also forget the last deck used by the Browser
        CardBrowser.clearLastDeckId()
        mFocusedDeck = did

        val deck = mDeckListAdapter.getNodeByDid(did)
        if (deck.hasCardsReadyToStudy()) {
            openReviewerOrStudyOptions(selectionType)
            return
        }
```

Reply:
I was a little frustrated with the additional latency on this call compared to when we were doing it in memory. 

This is what I had originally - more than happy to move back to it

Reply:
I get the feeling. I always loved micro-optimizations. Read dozens of times someone in StackOverflow telling that premature optimizations are the root of all evil.

Probably in the future we will be able to improve the background work perfomance when the repo's use of coroutines is more mature.